### PR TITLE
`CompileLoss`: Allow different but reconcilable structures for `y_true` and `y_pred`

### PR DESCRIPTION
**Pull request by nicolaspi**
Allow different but reconcilable structures for `y_true` and `y_pred`, e.g:
```python
y_true = (*, *)
y_pred = [*, *]
```
Fixes https://github.com/keras-team/keras/issues/20373

### CODE REVIEW

File: keras/src/trainers/compile_utils.py [hunk 1]
```diff
@@ -664,20 +665,20 @@ def call(self, y_true, y_pred, sample_weight=None):
...
-            ):
-                y_true = [y_true]
+            # y_true is either:
+            #   - flat or leaf
+            #   - has the same structure but uses different (but reconcilable)
+            #     container types, e.g `list` vs `tuple`
+            flat_y_true = tree.flatten(y_true)
             try:
-                y_true = tree.pack_sequence_as(y_pred, y_true)
+                y_true = tree.pack_sequence_as(y_pred, flat_y_true)
```
[the remark this answers is not in the document]
Comment (fchollet) [reply keras/src/trainers/compile_utils.py:674]: e.g. dicts vs lists...

Like, if you know you're looking specifically for tuples/lists conversion, you could hardcode that, maybe do a map to cast tuples to lists

[the remark this answers is not in the document]
Comment (nicolaspi) [reply keras/src/trainers/compile_utils.py]: > Like, if you know you're looking specifically for tuples/lists conversion, you could hardcode that, maybe do a map to cast tuples to lists

I made changes to do that.

**Comment by nicolaspi** [comment]
Putting it as a draft, as I have a better idea for solving this issue: make a function `assert_same_paths` that checks structures against the paths, independently from the containers' type.

**Comment by nicolaspi** [comment]
I made another PR: https://github.com/keras-team/keras/pull/20431, which this one depends on for completion.

**Review by fchollet — APPROVED** [review]
LGTM, thank you!

**Pull request merged**